[DO NOT MERGE] prepare for vfs 2.22.0

dscho opened the pull request:
This PR is set up to prepare for v2.22.0: its target branch tracks https://github.com/git-for-windows/git/pull/2192, rebasing VFS for Git's patches on top of it as frequently as possible.

Note that quite a few trace2 patches from VFS for Git were upstreamed (some in slightly different forms), so they are no longer visible in this PR's commits.

dscho commented:
Do we still use the `post-index-changed` hook? It was renamed to `post-index-change` when it was upstreamed...

dscho commented:
Hmm. It seems that `t/t5319-multi-pack-index.sh` [fails on Windows](https://gvfs.visualstudio.com/ci/_build/results?buildId=11751&view=logs&jobId=a02e0b12-7faa-5ff2-bb73-299918728bfa&taskId=54ad6422-a3c8-52b0-22c9-0d919ef00385&lineStart=3704&lineEnd=3705&colStart=1&colEnd=1) (also [in the slow PR build](https://gvfs.visualstudio.com/ci/_build/results?buildId=11750&view=logs&jobId=fd490c07-0b22-5182-fac9-6d67fe1e939b&taskId=ce91d5d6-0c55-50f5-8ab9-6695c03ab102&lineStart=7350&lineEnd=7351&colStart=1&colEnd=1)) with:

```
expecting success: 
	(
		cd dup &&
		git pack-objects --revs .git/objects/pack/pack-combined <<-EOF &&
		refs/heads/A
		^refs/heads/C
		EOF
		git multi-pack-index write &&
		ls .git/objects/pack | grep -v -e pack-[AB] >expect &&
		git multi-pack-index expire &&
		ls .git/objects/pack >actual &&
		test_cmp expect actual
	)

++ cd dup
++ git pack-objects --revs .git/objects/pack/pack-combined
3397c17c12e399df77c4f8ad89ef7ba8f628f947
++ git multi-pack-index write
++ ls .git/objects/pack
++ grep -v -e 'pack-[AB]'
++ git multi-pack-index expire
./test-lib.sh: line 908: 13244 Segmentation fault      git multi-pack-index expire
error: last command exited with $?=139
not ok 41 - expire removes unreferenced packs
```

@derrickstolee did you see this issue before?

dscho commented:
> ./test-lib.sh: line 908: 13244 Segmentation fault      git multi-pack-index expire

Turns out that this has been addressed in [v6 of `ds/midx-expire-repack`](https://github.com/gitgitgadget/git/pull/92). I papered over it by picking up the removal of `FREE_AND_NULL(m->packs[i]);` in `expire_midx_packs()` (after coming up with [a fix that removes the packs from the MRU](https://github.com/microsoft/git/pull/142)).

A much better idea is most likely to pick up v6 and drop the `expire` stuff that we have currently.

@derrickstolee do you want me to try my hand at updating this here PR with v6, or would you rather do it yourself?

derrickstolee commented:
@dscho I'll create a version of the v6 stuff that reverts the old code and plays the new code and make a PR against your `prepare` branch.

While you are re-rolling, I think it would be good to drop this merge and commit.

```
*   29d26a4035 Merge 'gvfs/ds/generation-numbers-update'
|\
| * 2478506054 commit: add generation to pop_most_recent_commit()
```
That `247850` adds a use of generation number, but only supplies `GENERATION_NUMBER_ZERO` to the callers, so is worthless.

derrickstolee commented:
We [got some more test failures in the full functional tests](https://github.com/microsoft/VFSForGit/pull/1209/checks?check_run_id=137193581) including this output difference:

```
status Output Lines counts do not match. was: 2 expected: 15
Extra: nothing to commit, working tree clean
Missing: Changes not staged for commit:
Missing:   (use "git add/rm <file>..." to update what will be committed)
Missing:   (use "git checkout -- <file>..." to discard changes in working directory)
Missing: 	deleted:    GVFS/GVFS/CommandLine/CloneHelper.cs
Missing: 	deleted:    GVFS/GVFS/CommandLine/CloneVerb.cs
Missing: 	deleted:    GVFS/GVFS/CommandLine/DiagnoseVerb.cs
Missing: 	deleted:    GVFS/GVFS/CommandLine/GVFSVerb.cs
Missing: 	deleted:    GVFS/GVFS/CommandLine/LogVerb.cs
Missing: 	deleted:    GVFS/GVFS/CommandLine/MountVerb.cs
Missing: 	deleted:    GVFS/GVFS/CommandLine/PrefetchHelper.cs
Missing: 	deleted:    GVFS/GVFS/CommandLine/PrefetchVerb.cs
Missing: 	deleted:    GVFS/GVFS/CommandLine/StatusVerb.cs
Missing: 	deleted:    GVFS/GVFS/CommandLine/UnmountVerb.
```

kewillford commented:
@derrickstolee as @dscho indicated the post index changed hook was renamed from `post-index-changed` to `post-index-change` when it was upstreamed so we will need to update the VFSForGit code or we will have failures since the projection will not be getting updated.

dscho commented:
> the post index changed hook was renamed from `post-index-changed` to `post-index-change` when it was upstreamed so we will need to update the VFSForGit code or we will have failures since the projection will not be getting updated.

I have an idea. How about special-casing that hook for now, and look for `.git/hooks/post-index-changed` if `...-change` was not found? That should work around things in the short run.

It might actually be necessary in the long run, too, unless we have a good upgrade story for already-cloned worktrees, to change the `...-changed` hook into a `...-change` one?

derrickstolee commented on Documentation/githooks.txt:
```diff
@@ -496,23 +496,25 @@ This hook is invoked by `git-p4 submit`. It takes no parameters and nothing
 from standard input. Exiting with non-zero status from this script prevent
 `git-p4 submit` from launching. Run `git-p4 submit --help` for details.
 
-post-index-change
```
Perhaps this is an incorrect merge-conflict resolution? I'm guessing we don't want to delete the section about the post-index-change hook.

dscho replied:
Good call!

dscho replied:
(And yes, this was a merge conflict resolution, I remember it... I had thought that there was already the correct section and I had removed the duplicate, incorrect section...)

dscho commented:
> I have an idea. How about special-casing that hook for now, and look for `.git/hooks/post-index-changed` if `...-change` was not found?

I made it so: 5b3b769e655a1557b54526c4793e80fb199ee804

@derrickstolee I also rebased after merging your #143, so the old midx-expire patches are gone.

dscho commented:
(@derrickstolee oh, and I put back the documentation for the `post-index-change` hook, thank you so much for your thorough review!)

dscho commented:
> > I have an idea. How about special-casing that hook for now, and look for `.git/hooks/post-index-changed` if `...-change` was not found?
> 
> I made it so: [5b3b769](https://github.com/microsoft/git/commit/5b3b769e655a1557b54526c4793e80fb199ee804)

Hmm. The test case that I introduced may be flaky. I'll need to look at this again tomorrow.

derrickstolee commented:
The failure in the "Windows - Extra" build is due to how I updated the multi-pack-index to add packs to the `packed_git` list. See microsoft/vfsforgit#1213 for full details. It does not require any Git fix.

derrickstolee commented:
Thanks for putting this together so quickly, @dscho! I think the current version looks great. I'll take another look after the full Git for Windows 2.22.0 release is final.

dscho commented:
> Thanks for putting this together so quickly, @dscho!

You're welcome. Thank you for upstreaming so many patches that made this round *so* much easier on me than, say, v2.20.0's.

> I think the current version looks great.

Yay!

> I'll take another look after the full Git for Windows 2.22.0 release is final.

Very good, thank you so much!

dscho commented on midx.c:
```diff
@@ -10,6 +10,7 @@
 #include "progress.h"
 #include "trace2.h"
 #include "run-command.h"
+#include "trace2.h"
```
Good point! ;-)

dscho commented:
Updated to v2.22.0-rc2.windows.1.

While at it, I dropped the `jk/no-sigpipe-during-network-transport` patches because I realized that they are already part of `v2.21.0`.

derrickstolee commented:
We may want to think about reverting [Ævar's change to not use the commit-graph while building the commit-graph](https://github.com/git/git/commit/43d356180556180b4ef6ac232a14498a5bb2b446). That will slow down our commit-graph writes significantly until the incremental file format is available. But I can also handle that in a follow-up PR.

dscho commented:
> But I can also handle that in a follow-up PR.

Please do: the changes you are talking about are not introduced by this PR (which tries to rebase VFSforGit's patches on top of v2.22.0-rc*) but by the upstream commits.

dscho commented:
For the record, I noticed that the old name of the `post-index-change` hook was `post-indexchanged`, not `post-index-changed`, so I changed that in the back-compat patch and force-pushed.

dscho commented:
Okay, y'all, I rebased onto Git for Windows v2.22.0 and force-pushed. This is it. Hopefully the Azure Pipeline succeeds without any help from my side, so that I can spend a couple of mostly offline days...

dscho commented:
> Hopefully the Azure Pipeline succeeds without any help from my side

Almost ;-) I had to retry once because of two flaky test cases that fail *sometimes* (and which are [on the backlog to be investigated](https://github.com/gitgitgadget/git/issues?utf8=%E2%9C%93&q=is%3Aissue+flaky+sort%3Aupdated-desc+)).

derrickstolee commented:
Thanks so much, @dscho! I'm going to build a PR build installer and run the VFS for Git tests one more time before updating `vfs-2.22.0` to the current `prepare-for-vfs-2.22.0` branch.
